src/hash_picker.cpp

```diff
@@ -0,0 +1,639 @@
+						, m_files.piece_length() / default_block_size);
+					ret.hash_failed[piece_index_t{pos.quot}].push_back(pos.rem);
+				}
+
+				m_merkle_trees[req.file][dest_start_idx + i] = tree[source_start_idx + i];
+			}
+			if (layer_size == 1) break;
+			dest_start_idx = merkle_get_parent(dest_start_idx);
+			source_start_idx = merkle_get_parent(source_start_idx);
+		}
+
+		// now copy the string of proof hashes
+		for (auto proof : proofs)
+		{
+			if (dest_start_idx & 1)
+			{
+				m_merkle_trees[req.file][dest_start_idx] = proof.first;
+				m_merkle_trees[req.file][dest_start_idx + 1] = proof.second;
+			}
+			else
+			{
+				m_merkle_trees[req.file][dest_start_idx - 1] = proof.first;
+				m_merkle_trees[req.file][dest_start_idx] = proof.second;
+			}
+			dest_start_idx = merkle_get_parent(dest_start_idx);
+		}
+
+		if (req.base == 0)
+		{
+			std::fill_n(m_hash_verified[req.file].begin() + req.index, req.count, true);
+			// TODO: add passing pieces to ret.hash_passed
+		}
+
+		if (req.base != 0)
+		{
+			TORRENT_ASSERT(req.base == m_piece_layer);
+			int const file_num_blocks = m_files.file_num_blocks(req.file);
+			int const file_first_leaf = m_files.file_first_block_node(req.file);
+			int const file_first_piece = m_files.file_first_piece_node(req.file);
+			int const base_layer_index = merkle_num_layers(merkle_num_leafs(file_num_blocks)) - req.base;
+			int const file_piece_offset = int(m_files.file_offset(req.file) / m_files.piece_length());
+
+			// it may now be possible to verify the hashes of previously received blocks
+			// try to verify as many child nodes of the received hashes as possible
+			for (int i = 0; i < req.count; ++i)
+			{
+				int const piece = req.index + i;
+				if (!m_merkle_trees[req.file][merkle_get_first_child(file_first_piece + piece)].is_all_zeros()
+					&& !m_merkle_trees[req.file][merkle_get_first_child(file_first_piece + piece) + 1].is_all_zeros())
+				{
+					// this piece is already verified
+					continue;
+				}
+
+				int const first_leaf = piece << req.base;
+				int const num_leafs = 1 << req.base;
+
+				bool done = false;
+				for (int j = 0; j < std::min(num_leafs, file_num_blocks - first_leaf); ++j)
+				{
+					if (m_merkle_trees[req.file][file_first_leaf + first_leaf + j].is_all_zeros())
+					{
+						done = true;
+						break;
+					}
+				}
+				if (done) continue;
+
+				merkle_fill_tree(m_merkle_trees[req.file], num_leafs, file_first_leaf + first_leaf);
+				if (m_merkle_trees[req.file][merkle_to_flat_index(base_layer_index, req.index + i)] != hashes[i])
+				{
+					merkle_clear_tree(m_merkle_trees[req.file], num_leafs / 2, merkle_get_parent(file_first_leaf + first_leaf));
+					m_merkle_trees[req.file][merkle_to_flat_index(base_layer_index, req.index + i)] = hashes[i];
+					TORRENT_ASSERT(num_leafs == m_files.piece_length() / default_block_size);
+					//verify_block_hashes(m_files.file_offset(req.file) / m_files.piece_length() + req.index);
+					// TODO: add to failed hashes
```

Comment:
this also sounds like an important thing to do

Reply:
this is the case where the block-hashes we got from a peer don't match what already know about the merkle tree. is that right?

Reply:
This is for the case were the client downloaded piece data before it knew the corresponding piece hashes. Now it has the hashes and discovered that some of the piece data was bad.